src/header/map.rs

```diff
@@ -445,12 +445,23 @@ impl HeaderMap {
     /// assert!(map.is_empty());
     /// assert_eq!(0, map.capacity());
     /// ```
+    #[inline]
     pub fn new() -> Self {
-        HeaderMap::try_with_capacity(0).unwrap()
+        Self::new_empty()
     }
 }
 
 impl<T> HeaderMap<T> {
+    fn new_empty() -> Self {
+        HeaderMap {
+            mask: 0,
+            indices: Box::new([]), // as a ZST, this doesn't actually allocate anything
+            entries: Vec::new(),
+            extra_values: Vec::new(),
+            danger: Danger::Green,
+        }
+    }
+
```

Comment:
```suggestion
        HeaderMap {
            mask: 0,
            indices: Box::new([]), // as a ZST, this doesn't actually allocate anything
            entries: Vec::new(),
            extra_values: Vec::new(),
            danger: Danger::Green,
        }
    }
}

impl<T> HeaderMap<T> {
```

I think it's simpler to make the `new` constructor initialize the empty map instead of adding a new method for it. `Vec::new` is implemented just like that, for example.

Reply:
That was my first inclination but it would mean we can't use `Self::new()` in the `Default` impl because `new()` is only defined for `HeaderMap<HeaderValue>`, not `HeaderMap<T>`. If we want to try to eliminate the extra method I'd suggest instead making `new()` and `try_with_capacity(0)` call `Default::default()` and putting the actual definition in there. Alternatively we can move the definition of `new()` to the more general impl block below but that might break calling code that is relying on `new()` to constrain the generic type.

Reply:
I went ahead and did this in the latest commit. After looking at https://doc.rust-lang.org/cargo/reference/semver.html#fn-generalize-compatible I wouldn't be opposed to just making `new()` more generic since that's considered a non-breaking change even if breaks type inference in consumers. I'm happy to do that here if that's what maintainers prefer.